packages/react-components/react-tree/src/components/Tree/useTree.ts

```diff
@@ -12,13 +22,168 @@ import type { TreeProps, TreeState } from './Tree.types';
+    onOpenChange?.(data.event, data);
+    if (!data.event.isDefaultPrevented()) {
+      setOpenSubtrees(updateOpenSubtrees(data, openSubtrees));
+    }
+  });
+  const { treeWalker: treeWalkerRef, root: treeRef } = useTreeWalker(NodeFilter.SHOW_ELEMENT, {
+    acceptNode: filterTreeItemAndSubtree,
+  });
+  const commonState = useSubtree(props, useMergedRefs(ref, treeRef));
+  return {
+    ...commonState,
+    openSubtrees,
+    requestOpenChange,
+    focusFirstSubtreeItem: useEventCallback(target => {
+      const treeWalker = treeWalkerRef.current;
+      if (!treeWalker) {
+        return;
+      }
+      const groupId = target.getAttribute('aria-owns');
+      if (groupId && targetDocument) {
+        const element = targetDocument.getElementById(groupId);
+        if (treeWalker && element) {
+          treeWalker.currentNode = element;
+          const firstTreeItem = treeWalker.firstChild() as TreeItemElement | null;
+          return firstTreeItem?.focus();
+        }
+      }
+    }),
+    focusSubtreeOwnerItem: useEventCallback(target => {
+      const treeWalker = treeWalkerRef.current;
+      if (!treeWalker) {
+        return;
+      }
+      treeWalker.currentNode = target;
+      const group = treeWalker.parentNode() as HTMLElement | null;
+      if (group) {
+        while (treeWalker.previousNode()) {
+          const treeItem = treeWalker.currentNode as TreeItemElement;
+          if (treeItem.getAttribute('aria-owns') === group.id) {
+            return treeItem.focus();
+          }
+        }
+      }
+    }),
+  };
+}
+
+function filterTreeItemAndSubtree(node: Node) {
+  const element = node as HTMLElement & { role: string };
+  return element.role === 'treeitem' || element.role === 'group' ? NodeFilter.FILTER_ACCEPT : NodeFilter.FILTER_REJECT;
+}
+
+function warnIfNoProperPropsSubtree(props: Pick<TreeProps, 'id' | 'aria-label' | 'aria-labelledby'>) {
+  if (process.env.NODE_ENV === 'development') {
+    if (!props.id) {
+      // eslint-disable-next-line no-console
+      console.warn('as sub Tree must have an id to be referred by a TreeItem');
+    }
+  }
+}
+
+function warnIfNoProperPropsRootTree(props: Pick<TreeProps, 'id' | 'aria-label' | 'aria-labelledby'>) {
+  if (process.env.NODE_ENV === 'development') {
+    if (!props['aria-label'] && !props['aria-labelledby']) {
+      // eslint-disable-next-line no-console
+      console.warn('Tree must have either a `aria-label` or `aria-labelledby` property defined');
+    }
+  }
+}
+
+function normalizeOpenSubtrees(openSubtrees?: string | string[]) {
+  if (!openSubtrees) {
+    return [];
+  }
+  return Array.isArray(openSubtrees) ? openSubtrees : [openSubtrees];
+}
+
+function normalizeOpenSubtreesOrUndefined(openSubtrees?: string | string[]) {
+  if (!openSubtrees) {
+    return undefined;
+  }
+  return normalizeOpenSubtrees(openSubtrees);
+}
+
+function updateOpenSubtrees(data: TreeOpenChangeData, previousOpenSubtrees: string[]) {
```

Comment:
not blocking, but have you considered using `Set`, I started using it for `Table` selection scenarios and honestly it makes updating things with ids so much cleaner

Reply:
Set is actually slower than `Array.include` methods for small sets of data, according to some benchmarks. I'm not sure I'd go for that, but we might analyze that in the future